algosdk/source_map.py

```diff
@@ -0,0 +1,97 @@
+from typing import Dict, Any, List, Tuple
+
+
+class SourceMap:
+    """
+    Decodes a VLQ-encoded source mapping between PC values and TEAL source code lines.
+    Spec available here: https://sourcemaps.info/spec.html
+
+    Args:
+        source_map (dict(str, Any)): source map JSON from algod
+        delimiter (str, optional): delimiter for mappings
+    """
+
+    def __init__(self, source_map: Dict[str, Any], delimiter: str = ";"):
+        self.delimter = delimiter
+
+        self.version: int = source_map["version"]
+        self.sources: List[str] = source_map["sources"]
+        self.mapping: str = source_map["mapping"]
+
+        pc_list = [
+            _decode_int_value(raw_val)
+            for raw_val in self.mapping.split(delimiter)
+        ]
+
+        # Initialize with 0,0 for pc/line
+        self.pc_to_line: Dict[int, int] = {0: 0}
+        self.line_to_pc: Dict[int, List[int]] = {0: [0]}
+
+        last_line = 0
+        for index, line_num in enumerate(pc_list):
+            if line_num is not None:  # be careful for '0' checks!
```

Comment:
No, the line_num may be None otherwise but I'm not sure exactly when. @algochoi ?

Reply:
IIRC we are converting from PC->line here, so consider a mapping like: `;1;;2` (I used ints rather than VLQ for readability).

Then `pc_list` would contain `{0:None, 1: 1, 2: None, 3: 2}` (the values are decoded from `_decode_int_value`). I think I left a comment in the original decoding scheme to remind myself that doing

```
if line_num:
    blah
```

could be misleading since a `0` for line_num could be valid.